### PR TITLE
Feature: Quiz Preview Screen

### PR DESCRIPTION
**Pull request by bharatchawla53**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,6 +14,7 @@
     "@types/react-dom": "^18.2.18",
     "axios": "^1.6.8",
     "bootstrap": "^5.3.2",
+    "date-fns": "^3.6.0",
```
Comment (omkarubale) [inline package.json:17]: We have `moment` installed for this. We do not need this package.

File: src/Kanbas/Courses/Quizzes/Preview/Question/Common/Header/index.css [hunk 2]
```diff
@@ -0,0 +1,15 @@
+.header {
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/Question/Common/Header/index.css:1]: Lets stay consistent with the class names in CSS and use a `wd-` prefix for custom class names.

Comment (omkarubale) [reply src/Kanbas/Courses/Quizzes/Preview/Question/Common/Header/index.css:1]: Also, use wd-quiz-question-header instead of just header, and follow this for other custom classes as well.

File: src/Kanbas/Courses/Quizzes/Preview/Question/Common/MultipleChoice/index.css [hunk 3]
```diff
@@ -0,0 +1,21 @@
+.text {
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/Question/Common/MultipleChoice/index.css:1]: Class name too generic. Lets use the `wd-quiz-question-text` and similar names.

Comment (omkarubale) [reply src/Kanbas/Courses/Quizzes/Preview/Question/Common/MultipleChoice/index.css:1]: These simple names can overlap with bootstrap

File: src/Kanbas/Courses/Quizzes/Preview/Question/Common/MultipleChoice/index.tsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import { eTrueOrFalse } from "../../../../../../store/enums/eTrueOrFalse";
+import "./index.css";
+
+function MultipleChoice({ question }: { question: any }) {
+
+    const renderAnswerOptions = () => {
+        if (question.quizQuestionType === eQuizQuestionType.MCQ) {
+            return renderChoices(question.choices);
+        } else if (question.quizQuestionType === eQuizQuestionType.TrueOrFalse) {
+            const eTrueOrFalseTypes = Object.keys(eTrueOrFalse).filter((key) => isNaN(Number(key)));
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/Question/Common/MultipleChoice/index.tsx:11]: Put this block in another method called `renderTrueFalseOptions()`.

File: src/Kanbas/Courses/Quizzes/Preview/Question/Common/ShortAnswers/index.tsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        <div className="text">
+            <div className="question-text">
+                <p>{question.questionText}</p>
+            </div>
+            <div className="answers">
+                {question.correctAnswers.map((choice: any, index: number) => (
+                    <div className="answer" key={index}>
+                        <label className="d-flex justify-content-start">
+                            <span className="answer-input">
+                                <input
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/Question/Common/ShortAnswers/index.tsx:12]: Use React-Bootstrap components. The one to use here is `Form.Control` and `Form.Label` along with `Form.Group`

File: src/Kanbas/Courses/Quizzes/Preview/Question/List/index.tsx [hunk 6]
```diff
@@ -1,6 +1,45 @@
...
+import { toggleFlagQuestion } from "../../../reducer";
+import { KanbasState } from "../../../../../store";
+import QuizQuestion from "..";
+
+function QuizQuestionPreviewList({ questionRefs }:
+    { questionRefs: React.MutableRefObject<(HTMLElement | null)[]> }) {
+
+    const dispatch = useDispatch();
+
+    const flaggedQuestions = useSelector(
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/Question/List/index.tsx:13]: This doesn't need to be in redux state, as this is only applicable for the quiz preview.
Instead, use another redux state for preview components or use `useState` and pass around the state via params.

File: src/Kanbas/Courses/Quizzes/Preview/Question/Single/index.tsx [hunk 7]
```diff
@@ -1,6 +1,63 @@
...
+import { KanbasState } from "../../../../../store";
+import { IKanbasQuizQuestion } from "../../../../../store/interfaces/quizzes";
+import { goToQuestion, toggleFlagQuestion } from "../../../reducer";
+import QuizQuestion from "..";
+import { MdArrowLeft, MdArrowRight } from "react-icons/md";
+
 function QuizQuestionPreviewSingle() {
+    const dispatch = useDispatch();
+
+    const flaggedQuestions = useSelector(
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/Question/Single/index.tsx:12]: This doesn't need to be in redux state, as this is only applicable for the quiz preview.
Instead, use another redux state for preview components or use useState and pass around the state via params.

File: src/Kanbas/Courses/Quizzes/Preview/Question/index.css [hunk 8]
```diff
@@ -0,0 +1,36 @@
+.question-holder {
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/Question/index.css:1]: Too generic names.

File: src/Kanbas/Courses/Quizzes/Preview/Question/index.tsx [hunk 9]
```diff
@@ -1,7 +1,63 @@
...
+function QuizQuestion({
+    question,
+    index,
+    flaggedQuestions,
+    handleFlagClick,
+    setRef }:
+    {
+        question: IKanbasQuizQuestion,
+        index: number,
+        flaggedQuestions: boolean[],
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/Question/index.tsx:19]: Why does a list of flagged questions need to be passed here? Instead, pass a method `isQuestionFlagged` from parent, and use that to get this info using parent's state.

File: src/Kanbas/Courses/Quizzes/Preview/Question/index.tsx [hunk 10]
```diff
@@ -1,7 +1,63 @@
...
+        handleFlagClick: (index: number) => void,
+        setRef?: (index: number, ref: HTMLDivElement | null) => void
+    }
+) {
+
+    const ref = useRef<HTMLDivElement>(null);
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/Question/index.tsx:25]: Why is this needed?

File: src/Kanbas/Courses/Quizzes/Preview/QuestionLinks/index.tsx [hunk 11]
```diff
@@ -1,6 +1,62 @@
...
+import { FaFlag } from "react-icons/fa";
+import { setActiveQuestion } from "../../reducer";
+
+function QuizPreviewQuestionLinks({ scrollToQuestion }:
+    { scrollToQuestion?: any }) {
+
+    const { courseId, quizId } = useParams();
+    const dispatch = useDispatch();
+
+    const flaggedQuestions = useSelector(
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/QuestionLinks/index.tsx:18]: This doesn't need to be in redux state, as this is only applicable for the quiz preview.
Instead, use another redux state for preview components or use useState and pass around the state via params.

File: src/Kanbas/Courses/Quizzes/Preview/index.css [hunk 12]
```diff
@@ -0,0 +1,6 @@
+.form-actions {
+    border: 1px solid #999;
+    margin: 18px 0 -1px;
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/index.css:3]: avoid negative margin

File: src/Kanbas/Courses/Quizzes/Preview/index.tsx [hunk 13]
```diff
@@ -5,26 +5,94 @@ import QuizQuestionPreviewSingle from "./Question/Single";
...
+                dispatch(setQuizQuestions(questions));
+                dispatch(resetPreview());
+            });
+        };
+    }, [dispatch]);
+
+    useEffect(() => {
+        const id = setInterval(() => {
+            setLastSavedTime(new Date());
+        }, 1000); // update every second
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/index.tsx:52]: We shouldn't need a timer for this. This should update every time a action goes through, like next, prev, etc or a state change occurs.

File: src/Kanbas/Courses/Quizzes/Preview/index.tsx [hunk 14]
```diff
@@ -5,26 +5,94 @@ import QuizQuestionPreviewSingle from "./Question/Single";
...
+import { resetPreview, setQuizQuestions } from "../reducer";
+import { getCurrentHumanReadableDate } from "../../common/Utils";
 
 function QuizPreview() {
 
-    // TODO this variable will be removed later to use props directly from JSON
-    const isQneQuestionAtATime = false;
+    const { quizId } = useParams();
+    const dispatch = useDispatch();
+    const questionRefs = useRef<(HTMLElement | null)[]>([]);
```
Comment (omkarubale) [inline src/Kanbas/Courses/Quizzes/Preview/index.tsx:24]: We dont need a ref of the HTML Element here. Instead, we can just pass around the state of questions, or using redux, rely on the specific question that is passed to the child component. I'd recommend using a preview specific redux state as this is not needed in the other areas of the app.

File: src/Kanbas/store/enums/eTrueOrFalse.ts [hunk 15]
```diff
@@ -0,0 +1,4 @@
+export enum eTrueOrFalse {
```
Comment (omkarubale) [inline src/Kanbas/store/enums/eTrueOrFalse.ts:1]: rename to `eBooleanChoices` and extend `eChoices`. We can keep the generic methods like the ones in src/Kanbas/Courses/Quizzes/Preview/Question/Common/MultipleChoice/index.tsx in `eChoices`.

**Review by omkarubale — CHANGES_REQUESTED** [review]
A few changes needed.

File: src/Kanbas/Courses/Quizzes/Preview/Question/Common/MultipleChoice/index.css [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+    margin-bottom: 1.5em;
+}
+
+.answers .answer {
+    padding: 5px;
+    margin: 0px;
+}
+
+.question.multiple-choice-question .answer {
+    border-top: 1px solid #ddd;
```
[the remark this answers is not in the document]
Comment (bharatchawla53) [reply src/Kanbas/Courses/Quizzes/Preview/Question/Common/MultipleChoice/index.css:16]: I meant to delete the other css styling since I did combine the logic in parent class.

File: src/Kanbas/Courses/Quizzes/Preview/Question/index.tsx (hunk 10 above)
Comment (bharatchawla53) [reply src/Kanbas/Courses/Quizzes/Preview/Question/index.tsx:25]: Removing it now since moved the logic to redux state.

File: src/Kanbas/store/enums/eTrueOrFalse.ts (hunk 15 above)
Comment (bharatchawla53) [reply src/Kanbas/store/enums/eTrueOrFalse.ts:1]: Typescript doesn't offer extend, unless you want me to create union of two enums to support that extends functionality?

**Comment by bharatchawla53** [comment]
Made the requested changes, and left some comments for unresolved conversations.

File: src/Kanbas/Courses/Quizzes/Preview/Question/List/index.tsx (hunk 6 above)
Comment (omkarubale) [reply src/Kanbas/Courses/Quizzes/Preview/Question/List/index.tsx:12]: This has not been addressed.

File: src/Kanbas/store/enums/eTrueOrFalse.ts (hunk 15 above)
Comment (omkarubale) [reply src/Kanbas/store/enums/eTrueOrFalse.ts]: No, that would overcomplicate it. Let's keep it as it is.

File: src/Kanbas/Courses/Quizzes/Preview/QuestionLinks/index.css [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+
+.wd-questions .list-group-item.active {
+    font-weight: bold;
+    background-color: transparent;
+}
+
+.wd-questions .list-group-item .flag-icon {
+    margin-bottom: 4px;
+    font-size: 10px;
+    margin-left: -15px;
```
[the remark this answers is not in the document]
Comment (omkarubale) [reply src/Kanbas/Courses/Quizzes/Preview/QuestionLinks/index.css:31]: We can come back to this after finishing up the main stuff.
